yarn-project/ethereum/src/l1_tx_utils.ts

```diff
@@ -0,0 +1,309 @@
+    private readonly logger?: DebugLogger,
+    config?: Partial<L1TxUtilsConfig>,
+  ) {
+    this.config = {
+      ...defaultL1TxUtilsConfig,
+      ...(config || {}),
+    };
+  }
+
+  /**
+   * Sends a transaction with gas estimation and pricing
+   * @param request - The transaction request (to, data, value)
+   * @param gasConfig - Optional gas configuration
+   * @returns The transaction hash and parameters used
+   */
+  public async sendTransaction(
+    request: L1TxRequest,
+    _gasConfig?: Partial<L1TxUtilsConfig>,
+  ): Promise<{ txHash: Hex; gasLimit: bigint; gasPrice: GasPrice }> {
+    const gasConfig = { ...this.config, ..._gasConfig };
+    const account = this.walletClient.account;
+
+    const gasLimit = await this.estimateGas(account, request);
+    const gasPrice = await this.getGasPrice(gasConfig);
+
+    const txHash = await this.walletClient.sendTransaction({
+      ...request,
+      gas: gasLimit,
+      maxFeePerGas: gasPrice.maxFeePerGas,
+      maxPriorityFeePerGas: gasPrice.maxPriorityFeePerGas,
+    });
+
+    this.logger?.verbose(
+      `Sent L1 transaction ${txHash} with gas limit ${gasLimit} and price ${formatGwei(gasPrice.maxFeePerGas)} gwei`,
+    );
+
+    return { txHash, gasLimit, gasPrice };
+  }
+
+  /**
+   * Monitors a transaction until completion, handling speed-ups if needed
+   * @param request - Original transaction request (needed for speed-ups)
+   * @param initialTxHash - Hash of the initial transaction
+   * @param params - Parameters used in the initial transaction
+   * @param gasConfig - Optional gas configuration
+   */
+  public async monitorTransaction(
+    request: L1TxRequest,
+    initialTxHash: Hex,
+    params: { gasLimit: bigint },
+    _gasConfig?: Partial<L1TxUtilsConfig>,
+  ): Promise<TransactionReceipt> {
+    const gasConfig = { ...this.config, ..._gasConfig };
+    const account = this.walletClient.account;
+
+    const tx = await this.publicClient.getTransaction({ hash: initialTxHash });
+    if (tx?.nonce === undefined || tx?.nonce === null) {
+      throw new Error(`Failed to get L1 transaction ${initialTxHash} nonce`);
+    }
+    const nonce = tx.nonce;
+
+    const txHashes = new Set<Hex>([initialTxHash]);
+    let currentTxHash = initialTxHash;
+    let attempts = 0;
+    let lastSeen = Date.now();
+
+    while (true) {
+      try {
+        const currentNonce = await this.publicClient.getTransactionCount({ address: account.address });
+        if (currentNonce > nonce) {
+          for (const hash of txHashes) {
+            try {
+              const receipt = await this.publicClient.getTransactionReceipt({ hash });
+              if (receipt) {
+                this.logger?.debug(`L1 Transaction ${hash} confirmed`);
+                if (receipt.status === 'reverted') {
+                  this.logger?.error(`L1 Transaction ${hash} reverted`);
+                  throw new Error(`Transaction ${hash} reverted`);
+                }
```

Comment:
This differs from viem's API, in that it does not throw if the tx reverts (I believe), so it may be unexpected to just throw. If do want to throw on revert, let's document it in the function jsdoc.